Fix `skipna` bug

bradyrx opened the pull request:
This PR fixes a bug with `skipna` where accurate results weren't reported if there were pairwise nans, e.g. nans that occurred in differing indices on `a` and `b`. I think this was only occurring with `pearson_r` and `spearman_r`.

The issue in these correlation ones was we would have `np.nanmean(a*b)` for instance, which dealt with nans in the same index. Then we would incorporate `np.nanmean(a*a)` which could bring in values that are actually nans in the `b` array. I hadn't thought of the pairwise issue since I usually work with model output that has nans in all the same places.

* Solves issue by creating nans at every pairwise location.
* Adds testing utilizing below case
* Adds testing for single time series for each metric compared to an outside package.
* Changed the name mad to median_absolute_error per #62. This is actually a median absolute error calculation and mae is already reserved for mean absolute error.
* Added tests to see that results from a gridded product match the same values as if the grid cells were fed in individually. Sometimes you get weird behavior with nans and vectorization, so it's worth the check.


Examples:

**Current master**:
```python
def drop_nans(a, b, dim="time"):
    """
    Masks a and b where they have pairwise nans.
    """
    a = a.where(b.notnull())
    b = b.where(a.notnull())
    return a.dropna(dim), b.dropna(dim)

time = pd.date_range("1/1/2000", "1/5/2000", freq="D")
a = xr.DataArray([3, np.nan, 5, 7, 9], dims=["time"], coords=[time])
b = xr.DataArray([7, 2, np.nan, 2, 4], dims=["time"], coords=[time])
a_dropped, b_dropped = drop_nans(a, b)

xs.pearson_r(a, b, "time", skipna=True)
>>> array(-0.58733524)
xs.pearson_r(a_dropped, b_dropped, "time")
>>> array(-0.73704347)

xs.spearman_r(a, b, "time", skipna=True)
>>> array(-0.42163702)
xs.spearman_r(a_dropped, b_dropped, "time")
>>> array(-0.5)

xs.mae(a, b, "time", skipna=True)
>>> array(4.66666667)
xs.mae(a_dropped, b_dropped, "time")
>>> array(4.66666667)
```

**With PR**:
```python
def drop_nans(a, b, dim="time"):
    """
    Masks a and b where they have pairwise nans.
    """
    a = a.where(b.notnull())
    b = b.where(a.notnull())
    return a.dropna(dim), b.dropna(dim)

time = pd.date_range("1/1/2000", "1/5/2000", freq="D")
a = xr.DataArray([3, np.nan, 5, 7, 9], dims=["time"], coords=[time])
b = xr.DataArray([7, 2, np.nan, 2, 4], dims=["time"], coords=[time])
a_dropped, b_dropped = drop_nans(a, b)

xs.pearson_r(a, b, "time", skipna=True)
>>> array(-0.73704347)
xs.pearson_r(a_dropped, b_dropped, "time")
>>> array(-0.73704347)

xs.spearman_r(a, b, "time", skipna=True)
>>> array(-0.5)
xs.spearman_r(a_dropped, b_dropped, "time")
>>> array(-0.5)

xs.mae(a, b, "time", skipna=True)
>>> array(4.66666667)
xs.mae(a_dropped, b_dropped, "time")
>>> array(4.66666667)
```

bradyrx commented:
CC @raybellwaves, @ahuang11, @hdrake, @aaronspring 

Not ready to merge due to the `pearson_r_p_value` and `spearman_r_p_value` issues right now. Also need to add accuracy testing for MAD but don't have a baseline now per https://github.com/raybellwaves/xskillscore/issues/62.

raybellwaves commented:
Thanks for working on this. Welcome to update the title as WIP:. Happy to take a look/brain storm when you're ready.

bradyrx commented on xskillscore/core/np_deterministic.py:
```diff
@@ -127,15 +150,18 @@ def _pearson_r_p_value(a, b, weights, axis, skipna):
         a = np.rollaxis(a, axis)
         b = np.rollaxis(b, axis)
-        dof = np.apply_over_axes(np.sum, np.isnan(a * b), 0).squeeze() - 2
-        dof = np.where(dof > 1.0, dof, a.shape[0] - 2)
```
These two lines were confusing to me and were breaking cases with nans. The below code `np.count_nonzero` is a straight forward way to get at the degrees of freedom.

I imagine line 138 was to deal with cases with 0 dof? But it replaced 0 or less dof with the max dof in this case? I don't fully understand, but it seemed to work anyways. So with the method below we don't get `dof < 1` cases to return `1.0` for the p-value like in `scipy.stats.pearsonr`, but instead it just returns a nan.

bradyrx commented:
Okay @raybellwaves, this is ready to be looked at.

bradyrx commented:
A bit of scope creep here, but did the following:

* Changed the name `mad` to `median_absolute_error` per https://github.com/raybellwaves/xskillscore/issues/62. This is actually a median absolute error calculation and `mae` is already reserved for mean absolute error.

* Added tests to see that results from a gridded product match the same values as if the grid cells were fed in individually. Sometimes you get weird behavior with nans and vectorization, so it's worth the check.

Testing suite still runs in ~7 seconds. So although some of my tests might be overkill, it's reasonable to run on PRs, etc.

bradyrx commented on CHANGELOG.rst:
```diff
@@ -0,0 +1,25 @@
+=================
+Changelog History
```
Previous versions can be added retroactively. I think this could be done pretty easily in a single PR since there aren't too many PRs historically here. This will be nice to have once we have the `sphinx` documentation going.

hdrake commented:
Just skimmed the thread above but all looks good to me.

raybellwaves commented:
LGTM. Just had some minor tweaks to use similar variables in the testing.

bradyrx commented:
Sounds good to me @raybellwaves. We can work on https://github.com/raybellwaves/xskillscore/issues/64 following this merge. I don't have time to lead that one right now though.

raybellwaves commented:
@bradyrx if you are happy mind merging https://github.com/bradyrx/xskillscore/pull/1

I wanted to do a PR rather that push as to not override your PR. I tested my branch so should be good.

I'll then merge this one.

bradyrx commented:
Oops didn't see that. Just merged -- should be good to go following CI.

raybellwaves commented:
Thanks @bradyrx
